locator/topology.cc

```diff
@@ -580,20 +580,31 @@ void topology::sort_by_proximity(locator::host_id address, host_id_vector_replic
 
 template <typename T>
 void topology::do_sort_by_proximity(T address, utils::small_vector<T, 3>& addresses) const {
-    std::sort(addresses.begin(), addresses.end(), [this, &address](T& a1, T& a2) {
-        return compare_endpoints(address, a1, a2) < 0;
+    struct info {
+        T id;
+        int distance;
+    };
+    utils::small_vector<info, 3> host_infos;
+    host_infos.reserve(addresses.size());
+    const auto& loc = get_location(address);
+    for (const auto& id : addresses) {
+        const auto& loc1 = get_location(id);
+        host_infos.emplace_back(id, distance(address, loc, id, loc1));
+    }
```

Comment:
Using std::range transformations and std::ranges::to would be faster since it can use the from_range_t constructor which avoids checks after every emplace_back (which the compiler may not be able to eliminate).

Reply:
The distance calculations could be done during effective_replication_map preparation. In fact we could prepare all the shuffled variants and pick one randomly, for small replication factors. But out of scope for now.

Reply:
good idea, will do

Reply:
> The distance calculations could be done during effective_replication_map preparation. In fact we could prepare all the shuffled variants and pick one randomly, for small replication factors. But out of scope for now.

But the distance is from the coordinator node? Would we prepare the distance vectors against each possible node in the system?